File: qadevOOo/tests/java/ifc/sheet/_XSheetFilterableEx.java
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */
package ifc.sheet;

import lib.MultiMethodTest;

import com.sun.star.sheet.TableFilterField;
import com.sun.star.sheet.XSheetFilterDescriptor;
import com.sun.star.sheet.XSheetFilterable;
import com.sun.star.sheet.XSheetFilterableEx;
import com.sun.star.uno.UnoRuntime;


/**
 * Interface test for XSheetFilterableEx
 *
 * methods:
 *
 *  createFilterDescriptorByObject
 *
 */
public class _XSheetFilterableEx extends MultiMethodTest {
    public XSheetFilterableEx oObj = null;

    /**
     * queries a XSheetFilterable from the parent Object and then creates
     * a XSheetFilterDescriptor with it.
     * Returns OK if the Descriptor isn't null and checkFilterDescriptor returns true
     */

    public void _createFilterDescriptorByObject() {
        boolean res = true;
        Object parent = tEnv.getTestObject();
        XSheetFilterable aFilterable = UnoRuntime.queryInterface(
                                               XSheetFilterable.class, parent);
        XSheetFilterDescriptor desc = oObj.createFilterDescriptorByObject(
                                              aFilterable);

        if (desc != null) {
            res &= checkFilterDescriptor(desc);
        } else {
            log.println("gained XSheetFilterDescriptor is null");
            res &= false;
        }
        tRes.tested("createFilterDescriptorByObject()",res);
    }

    /**
     * returns true if the Method getFilterFields() returns a non empty array
     * and all Fields can be gained without exception
     */

    public boolean checkFilterDescriptor(XSheetFilterDescriptor desc) {
        TableFilterField[] fields = desc.getFilterFields();
        boolean res = true;

        if (fields.length == 0) {
            log.println("The gained Descriptor is empty");
            res &= false;
        } else {
               log.println("Found "+fields.length+" TableFields");
        }

        for (int k = 0; k < fields.length; k++) {
            log.println("StringValue(" + k + "): " + fields[k].StringValue);
            log.println("IsNumeric(" + k + "): " + fields[k].IsNumeric);
            log.println("NumericValue(" + k + "): " + fields[k].NumericValue);
            log.println("Field(" + k + "): " + fields[k].Field);
            log.println("Connection(" + k + "): " + fields[k].Connection);
            log.println("Operator(" + k + "): " + fields[k].Operator);
        }

        return res;
    }
}

File: test/source/sheet/xsheetfilterableex.cxx
/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4; fill-column: 100 -*- */
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include <test/sheet/xsheetfilterableex.hxx>

#include <com/sun/star/sheet/FilterConnection.hpp>
#include <com/sun/star/sheet/FilterOperator.hpp>
#include <com/sun/star/sheet/TableFilterField.hpp>
#include <com/sun/star/sheet/XSheetFilterDescriptor.hpp>
#include <com/sun/star/sheet/XSheetFilterable.hpp>
#include <com/sun/star/sheet/XSheetFilterableEx.hpp>
#include <com/sun/star/sheet/XSpreadsheet.hpp>
#include <com/sun/star/table/XCellCursor.hpp>
#include <com/sun/star/uno/Reference.hxx>
#include <com/sun/star/uno/Sequence.hxx>

#include <cppunit/extensions/HelperMacros.h>

using namespace css;
using namespace css::uno;

namespace apitest
{
void XSheetFilterableEx::testCreateFilterDescriptorByObject()
{
    uno::Reference<sheet::XSheetFilterableEx> xSFEx(init(), UNO_QUERY_THROW);
    uno::Reference<sheet::XSheetFilterable> xSF(getXSpreadsheet(), UNO_QUERY_THROW);

    uno::Reference<sheet::XSheetFilterDescriptor> xSFD = xSFEx->createFilterDescriptorByObject(xSF);
    CPPUNIT_ASSERT_MESSAGE("no XSheetFilterDescriptor", xSFD.is());

    uno::Sequence<sheet::TableFilterField> xTFF = xSFD->getFilterFields();
    CPPUNIT_ASSERT_MESSAGE("The gained XSheetFilterDescriptor is empty", xTFF.getLength() != 0);

    for (const auto& field : xTFF)
    {
        // we don't care about the actual value, just that we can access the fields
        CPPUNIT_ASSERT_MESSAGE("Unable to retrieve field: StringValue",
                               !field.StringValue.isEmpty() || field.StringValue.isEmpty());
        CPPUNIT_ASSERT_MESSAGE("Unable to retrieve field: IsNumeric",
                               !field.IsNumeric || field.IsNumeric);
        CPPUNIT_ASSERT_MESSAGE("Unable to retrieve field: NumericValue",
                               field.NumericValue != 0.0 || field.NumericValue == 0.0);
        CPPUNIT_ASSERT_MESSAGE("Unable to retrieve field: Field",
                               field.Field != 0 || field.Field == 0);
        CPPUNIT_ASSERT_MESSAGE("Unable to retrieve field: Connection",
                               field.Connection == sheet::FilterConnection_AND
                                   || field.Connection == sheet::FilterConnection_OR);
        CPPUNIT_ASSERT_MESSAGE("Unable to retrieve field: Operator",
                               field.Operator != sheet::FilterOperator_EMPTY);
    }
}
}

/* vim:set shiftwidth=4 softtabstop=4 expandtab cinoptions=b1,g0,N-s cinkeys+=0=break: */
